Add customizable user roles

Gargron opened the pull request:
This replaces hardcoded `admin` and `moderator` roles that have predefined permission sets with a customizable user role system. Roles can be created with various permission flags for individual features and areas of Mastodon.

The base role "Everyone" is inherited by every signed-in user without an assigned role as well as all other roles. Currently, the only *practical* setting in the "Everyone" role is "Invite users", which replaces the "Allow invitations by" site setting.

Default user roles are:

- "Moderator"
- "Admin"
- "Owner"

Example role ideas:

- "Analyst", with only access to "View Dashboard", allowing retrieval of various metrics
- "Auditor", with only access to "View Audit Log", allowing the perusal of the log of administrative actions
- "Content reviewer", with only access to "Manage Taxonomies", allowing the review of trending content

## Permission flags

|Flag|Value|Description|
|--|--|--|
|Administrator|0x0000000000000001|Users with this permission will bypass every permission|
|Devops|0x0000000000000002|Allows users to access Sidekiq and pgHero dashboards|
|View Audit Log|0x0000000000000004|Allows users to see a history of administrative actions on the server|
|View Dashboard|0x0000000000000008|Allows users to access the dashboard and various metrics|
|Manage Reports|0x0000000000000010|Allows users to review reports and perform moderation actions against them|
|Manage Federation|0x0000000000000020|Allows users to block or allow federation with other domains, and control deliverability|
|Manage Settings|0x0000000000000040|Allows users to change site settings|
|Manage Blocks|0x0000000000000080|Allows users to block e-mail providers and IP addresses|
|Manage Taxonomies|0x0000000000000100|Allows users to review trending content and update hashtag settings|
|Manage Appeals|0x0000000000000200|Allows users to review appeals against moderation actions|
|Manage Users|0x0000000000000400|Allows users to view other users' details and perform moderation actions against them|
|Manage Invites|0x0000000000000800|Allows users to browse and deactivate invite links|
|Manage Rules|0x0000000000001000|Allows users to change server rules|
|Manage Announcements|0x0000000000002000|Allows users to manage announcements on the server|
|Manage Custom Emojis|0x0000000000004000|Allows users to manage custom emojis on the server|
|Manage Webhooks|0x0000000000008000|Allows users to set up webhooks for administrative events|
|Invite Users|0x0000000000010000|Allows users to invite new people to the server|
|Manage Roles|0x0000000000020000|Allows users to manage and assign roles below theirs|
|Manage User Access|0x0000000000040000|Allows users to disable other users' two-factor authentication, change their e-mail address, and reset their password|
|Delete User Data|0x0000000000080000|Allows users to delete other users' data without delay|

## API changes

The response of `GET /api/v1/verify_credentials` gets a new attribute, `role`, containing the `id`, `name`, `permissions`, `highlighted` and `color` attributes. The `permissions` string is a bitmask that can be compared against known flags to present the end-user with appropriate UI controls in apps.

## Screenshots

**List of roles**

![image](https://user-images.githubusercontent.com/184731/175862843-61362b42-6484-4cb8-ab50-c0c48133ee33.png)  

**Editing the base role**

![image](https://user-images.githubusercontent.com/184731/175862913-8acf9372-d113-472b-b55a-0846cba20500.png)  

**Editing a normal role**

![image](https://user-images.githubusercontent.com/184731/175863018-6c7270d9-73e3-4554-9ee3-8bacc38a67aa.png) 

**Filtering users by role**

![image](https://user-images.githubusercontent.com/184731/175863070-a3e46445-09ff-48b7-bdb2-0a357a67f6a6.png) 

**Assigning a role to a user**

![image](https://user-images.githubusercontent.com/184731/175863124-e84dedb2-7517-487d-8d21-6ecbb10b7eb9.png) 

___

This project was funded through the [NGI0 Discovery](https://nlnet.nl/discovery) Fund, a fund established by [NLnet](https://nlnet.nl/) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu/) programme, under the aegis of DG Communications Networks, Content and Technology under grant agreement No 825322.

Gargron committed:
Add customizable user roles

ClearlyClaire commented:
![image](https://user-images.githubusercontent.com/384364/174808224-89b83ce6-37e0-492b-97de-77278ef03d12.png)

I will review this, but honestly, even if I understand that a lot of this is pretty intertwined, this being a single-PR single-commit ordeal changing 173 files is vastly unreasonable. There is no reason this couldn't have been broken down in more manageable pieces. It is not [the first time](https://github.com/mastodon/mastodon/pull/17431#pullrequestreview-880508907) I bring that up, and it's pretty discouraging that it hasn't changed.

ClearlyClaire commented on app/views/admin/accounts/index.html.haml:
```diff
@@ -4,45 +4,36 @@
-    - (AccountFilter::KEYS - %i(origin status permissions)).each do |key|
-      - if params[key].present?
-        = hidden_field_tag key, params[key]
+  .filters
+    .filter-subset.filter-subset--with-select
+      %strong= t('admin.accounts.location.title')
+      .input.select.optional
+        = select_tag :origin, options_for_select([[t('admin.accounts.location.local'), 'local'], [t('admin.accounts.location.remote'), 'remote']], params[:origin]), prompt: I18n.t('generic.all')
+    .filter-subset.filter-subset--with-select
+      %strong= t('admin.accounts.moderation.title')
+      .input.select.optional
+        = select_tag :status, options_for_select([[t('admin.accounts.moderation.active'), 'active'], [t('admin.accounts.moderation.silenced'), 'silenced'], [t('admin.accounts.moderation.suspended'), 'suspended'], [safe_join([t('admin.accounts.moderation.pending'), "(#{number_with_delimiter(User.pending.count)})"], ' '), 'pending']], params[:status]), prompt: I18n.t('generic.all')
+    .filter-subset.filter-subset--with-select
+      %strong= t('admin.accounts.role')
+      .input.select.optional
+        = select_tag :role_id, options_from_collection_for_select(UserRole.assignable, :id, :name, params[:role_id]), prompt: I18n.t('admin.accounts.moderation.all')
+    .filter-subset.filter-subset--with-select
+      %strong= t 'generic.order_by'
+      .input.select
+        = select_tag :order, options_for_select([[t('relationships.most_recent'), nil], [t('relationships.last_active'), 'active']], params[:order])
```
I admit I haven't actually run the code yet so I'm unsure what the admin interface looks like in terms of graphical design, especially with the select tags. It would have been nice to have screenshots in the PR's description.

Additionally, the select tags may require a few more clicks in some scenarios. But I guess they would also allow a few less clicks, and more importantly, client-server back-and-forth in some other scenarios, so I don't know what to think of it exactly.

Gargron replied:
Using a select tag for the roles was the only way to avoid the filters breaking into multiple lines. And select tags unfortunately do not look good next to the tabs, so I had to change everything into select tags.

ClearlyClaire reviewed (changes requested):
Once again, on the PR process itself, this should have at least been split in smaller commits. It could have easily been done so by, for instance, starting with two permissions equivalent to the existing `admin` and `moderator` roles, then gradually splitting them in other permissions. The `manage_roles` especially is more complex, involving security properties and its full implementation requiring new endpoints. The `invite_users` role also changed things quite a bit, with moving away from some global setting. Anyway, what is done is done, but I hope you have more consideration for your would-be reviewers in the future.

The code itself looks good overall, but there are a some important bugs to iron out before merging. I have left plenty of inline comments, but here are the more general and prevalent points.

There are discrepancies in handling the `administrator` permission. Unless I am mistaken, in some cases it bypasses other permissions, in others it doesn't. And if I'm not mistaken, it's set in such a way that existing admins… won't actually have any permissions besides `administrator`? Which would prevent them from creating roles with more granular permissions, and would exempt them from some notifications and checks.

This is seriously lacking in tests: while, the existing tests have been modified to test the default Administrator and Moderator roles that somewhat mimic the previous roles, there is no test at all for more granular permissions. The new endpoints, models and class methods (e.g. `User.those_who_can`) have no tests whatsoever either.

The roles are not translatable, which may be an issue for servers with admins of different languages. It's also an issue with the preset roles when the server admins are not English-speaking.

The whole “position” thing is confusing and poorly explained. As far as I understand, it's an arbitrary number that defines in which order roles are displayed on the role management interface, as well as ensure you can only demote/suspend/etc. users with a lower “role” than yours. The only user facing text describing that setting, “Higher role decides conflict resolution in certain situations” does not come close to explaining that.

Furthermore, I believe there are four issues that open the system to privilege escalation attacks for users with the `manage_roles` permission and access to another account with a lower role:
- you can create roles with a higher position than yours
- you can edit your own role to bump its position
- you can change the role of any user with a lower position to a role with a higher position than yours
- although you can't create roles with more permissions than yours, you can change the role of any user with a lower position to a role with more permissions than you

Finally, it seems possible to remove permissions from your current role as well as lower its position. I don't know if that's a bad thing or not, to be honest, but that seems like something with which you could end up shooting yourself in the foot.

Gargron commented on app/controllers/admin/dashboard_controller.rb:
```diff
@@ -5,7 +5,9 @@ class DashboardController < BaseController
-      @system_checks         = Admin::SystemCheck.perform
+      authorize :dashboard, :index?
+
+      @system_checks         = Admin::SystemCheck.perform(current_user)
```
I think it's the right thing to do. Technical details should not be exposed except to those with the devops permission. And when server rules are missing, the warning links to "Add server rules", but that is useless unless you have the permission to add server rules.

Gargron commented on app/policies/rule_policy.rb:
```diff
@@ -2,18 +2,18 @@
 class RulePolicy < ApplicationPolicy
   def index?
-    staff?
+    role.can?(:manage_rules)
```
Those who have the `manage_reports` permission can also see the rules on the report view, plus on the public page. Given that rules are public, it doesn't make sense to add a `view_rules` flag, since everyone should have it then, but then everyone would have access to one administration menu item.

ClearlyClaire replied:
hm… I'm not sure about it, but it's not a big deal and can always be addressed later

Gargron commented on config/navigation.rb:
```diff
@@ -2,66 +2,67 @@
 SimpleNavigation::Configuration.run do |navigation|
   navigation.items do |n|
-    n.item :web, safe_join([fa_icon('chevron-left fw'), t('settings.back')]), root_url
+    n.item :web, safe_join([fa_icon('chevron-left fw'), t('settings.back')]), root_path
```
Consistency, since it was (for some reason) a mix of both

Gargron commented on config/navigation.rb:
```diff
@@ -2,66 +2,67 @@
     end
 
-    n.item :logout, safe_join([fa_icon('sign-out fw'), t('auth.logout')]), destroy_user_session_url, link_html: { 'data-method' => 'delete' }
+    n.item :sidekiq, safe_join([fa_icon('diamond fw'), 'Sidekiq']), sidekiq_path, link_html: { target: 'sidekiq' }, if: -> { current_user.can?(:view_devops) }
+    n.item :pghero, safe_join([fa_icon('database fw'), 'PgHero']), pghero_path, link_html: { target: 'pghero' }, if: -> { current_user.can?(:view_devops) }
```
If all you have is the `view_devops` permission, there is no way to access them if they are under the administration menu, since they open outside the admin UI.

Gargron commented on app/services/notify_service.rb:
```diff
@@ -76,7 +76,7 @@ def response_to_recipient?
 
   def from_staff?
-    @notification.from_account.local? && @notification.from_account.user.present? && @notification.from_account.user.staff?
+    @notification.from_account.local? && @notification.from_account.user.present? && @notification.from_account.user_role&.overrides?(@recipient.user_role)
```
Since there is no concept of "staff" anymore (and I can imagine roles being used for stuff like "patreon supporter" etc without any staff-like implications), using role priority is the only way I can see to re-implement the intended effect here.

Gargron committed:
Various fixes and improvements

Gargron commented on app/policies/account_policy.rb:
```diff
@@ -2,74 +2,66 @@
 
   def destroy?
-    record.suspended_temporarily? && admin?
+    record.suspended_temporarily? && role.can?(:manage_users)
```
Created new permission `delete_user_data`

Gargron commented on app/policies/account_policy.rb:
```diff
@@ -2,74 +2,66 @@
 
   def memorialize?
-    admin? && !record.user&.admin? && !record.instance_actor?
+    role.can?(:manage_users) && role.overrides?(record.user_role) && !record.instance_actor?
```
Assigned this to new permission `delete_user_data`

Gargron commented on app/controllers/api/v1/admin/accounts_controller.rb:
```diff
@@ -119,7 +119,7 @@ def translated_filter_params
     end
 
-    translated_params[:permissions] = 'staff' if params[:staff].present?
+    translated_params[:permissions] = 'staff' if params[:staff].present? # FIXME
```
Added translation layer

Gargron commented on app/models/account_filter.rb:
```diff
@@ -4,7 +4,7 @@ class AccountFilter
     origin
     status
-    permissions
+    role_id
```
Added translation layer

Gargron commented on app/views/custom_css/show.css.erb:
```diff
@@ -0,0 +1,10 @@
+  --user-role-accent: <%= role.color %>;
+}
+
+<%- end %>
```
Perhaps it's no longer named appropriately but it is the dynamically generated CSS we already load

Gargron commented on app/models/user_role.rb:
```diff
@@ -0,0 +1,107 @@
+    manage_federation: (1 << 5),
+    manage_settings: (1 << 6),
+    manage_blocks: (1 << 7),
+    manage_taxonomies: (1 << 8),
```
Taxonomy is what both tags and categories were named in WordPress, so I picked it as the overarching term for stuff that refers to other stuff - hashtags, recommendations, and trending posts and links.

ClearlyClaire replied:
ok but that does still sound weird and unintuitive to me :/

Gargron commented on app/models/user.rb:
```diff
@@ -135,9 +137,29 @@ class User < ApplicationRecord
   attr_reader :invite_code
   attr_writer :external, :bypass_invite_request_check
 
+  def self.those_who_can(*privileges)
```
Made more clear now

Gargron commented on app/models/user_role.rb:
```diff
@@ -0,0 +1,107 @@
+  end
+
+  def validate_permissions_elevation
+    errors.add(:permissions_as_keys, :elevated) if defined?(@current_account) && @current_account.user_role.permissions & permissions != permissions
```
Addressed

Gargron commented on app/models/user.rb:
```diff
@@ -135,9 +137,29 @@ class User < ApplicationRecord
   attr_writer :external, :bypass_invite_request_check
 
+  def self.those_who_can(*privileges)
+    matching_role_ids = UserRole.where('(permissions & ?) != 0', privileges.reduce(0) { |mask, privilege| mask | UserRole::FLAGS[privilege] }).pluck(:id)
```
Addressed

Gargron commented on db/seeds/03_roles.rb:
```diff
@@ -0,0 +1,4 @@
+everyone  = UserRole.everyone
+moderator = UserRole.create_with(highlighted: true, permissions: UserRole::FLAGS[:view_dashboard] | UserRole::FLAGS[:view_audit_log] | UserRole::FLAGS[:manage_users] | UserRole::FLAGS[:manage_reports] | UserRole::FLAGS[:manage_taxonomies], position: 10).find_or_create_by(name: 'Moderator')
+admin     = UserRole.create_with(highlighted: true, permissions: moderator.permissions | UserRole::FLAGS[:manage_federation] | UserRole::FLAGS[:manage_settings] | UserRole::FLAGS[:manage_blocks] | UserRole::FLAGS[:manage_appeals] | UserRole::FLAGS[:manage_rules] | UserRole::FLAGS[:manage_invites] | UserRole::FLAGS[:manage_announcements] | UserRole::FLAGS[:manage_custom_emojis] | UserRole::FLAGS[: ...
+owner     = UserRole.create_with(highlighted: true, permissions: UserRole::FLAGS[:administrator], position: 1000).find_or_create_by(name: 'Owner')
```
Addressed

Gargron commented on spec/controllers/admin/accounts_controller_spec.rb:
```diff
@@ -6,7 +6,7 @@
 
   describe 'GET #index' do
-    let(:current_user) { Fabricate(:user, admin: true) }
+    let(:current_user) { Fabricate(:user, role: UserRole.find_by(name: 'Admin')) }
```
My intention was not to rewrite every old test, but to ensure everything worked more or less as before

Gargron commented on db/post_migrate/20220617202502_migrate_roles.rb:
```diff
@@ -0,0 +1,26 @@
+    moderator_role = UserRole.find_by(name: 'Moderator')
+
+    User.where(role_id: owner_role.id).in_batches.update_all(admin: true)
+    User.where(role_id: moderator_role.id).in_batches.update_all(moderator: true)
```
Addressed

Gargron commented on app/policies/user_policy.rb:
```diff
@@ -2,52 +2,38 @@
+  end
+
+  def change_role?
+    role.can?(:manage_roles) && role.overrides?(record.role)
```
Addressed

Gargron commented on app/controllers/admin/users/roles_controller.rb:
```diff
@@ -0,0 +1,31 @@
+    def update
+      authorize @user, :change_role?
+
+      if @user.update(resource_params)
```
Addressed

Gargron commented on app/views/admin/roles/_form.html.haml:
```diff
@@ -0,0 +1,30 @@
+    %hr.spacer/
+
+  .field-group
+    = f.input :permissions_as_keys, collection: UserRole::FLAGS.keys, wrapper: :with_block_label, include_blank: false, label_method: lambda { |privilege| safe_join([t("admin.roles.privileges.#{privilege}"), content_tag(:span, t("admin.roles.privileges.#{privilege}_description"), class: 'hint')]) }, required: false, as: :check_boxes, collection_wrapper_tag: 'ul', item_wrapper_tag: 'li'
```
Addressed

Gargron commented on app/controllers/api/v1/admin/measures_controller.rb:
```diff
@@ -1,11 +1,15 @@
+  after_action :verify_authorized
+
   def create
+    authorize :dashboard, :index?
```
Addressed by hiding the counters when the permission is missing

ClearlyClaire replied:
It seems a bit weird to me to allow managing federation without allowing to see those things, but it's not a huge deal.

Gargron commented on app/controllers/api/v1/admin/trends/tags_controller.rb:
```diff
@@ -1,11 +1,15 @@
+  after_action :verify_authorized
+
   def index
+    authorize :dashboard, :index?
```
This API is used to display trends on the dashboard. I've decided to address this by making the API return public results when the user doesn't have the `manage_taxonomies` permission instead of throwing a HTTP 403.

Gargron commented on app/controllers/api/v1/admin/trends/statuses_controller.rb:
```diff
@@ -1,11 +1,15 @@
+  after_action :verify_authorized
+
   def index
+    authorize :dashboard, :index?
```
This API will now mirror `GET /api/v1/trends/statuses` when accessed without `manage_taxonomies` permission instead of throwing a HTTP 403.

Gargron commented on app/controllers/api/v1/admin/trends/links_controller.rb:
```diff
@@ -1,11 +1,15 @@
+  after_action :verify_authorized
+
   def index
+    authorize :dashboard, :index?
```
This API will now mirror `GET /api/v1/trends/statuses` when accessed without `manage_taxonomies` permission instead of throwing a HTTP 403.

Gargron commented:
Updated according to feedback:

- Add visual separation in categories of permissions in form
- Change admin trends APIs to mirror public trends APIs when accessed without `manage_taxonomies` permission
- Change deleting and memorializing accounts to be behind `delete_user_data` permission
- Change resetting password, changing e-mail and disabling 2FA to be behind `manage_user_access` permission
- Fix being able to assign a role with a higher position than yours
- Fix being able to create a role with a higher position than yours
- Fix backwards compatibility with `staff` and `permissions` filter params in admin API
- Fix form displaying all permissions for the "everyone" role
- Fix counters being displayed on various pages if user doesn't have `view_dashboard` permission
- Fix `role_id` index on users table including null rows

Gargron committed:
Merge remote-tracking branch 'origin/main' into feature-dynamic-roles

ClearlyClaire commented on app/models/user_role.rb:
```diff
@@ -0,0 +1,179 @@
+
+    # Otherwise, compute permissions based on special conditions
+    @computed_permissions ||= begin
+      permissions = self.class.everyone.permissions | self.permissions
```
I didn't catch that before, but couldn't it make sense to have a role that has *less* permissions than `everyone`? e.g. everyone has by default the right to invite people, but someone can lose that right

Gargron replied:
The way to do that would be to move privileges to another role that *isn't* `everyone` and then grant that role to people except those who lost the right. Of course you could also make a role have an option to not inherit from everyone, but I'd leave that until someone asks for it.

ClearlyClaire reviewed (changes requested):
A lot of improvements overall, but I think the explanation for the “position” stuff is still inadequate. Localization concern are still a thing as well.

In addition, there are a few issues I hadn't noticed in my previous review:
- with the migration script, normal users will gain the ability to send invites, even on instances which had them disabled
- the same is true for the `show_staff_badge` setting
- ideally, the permission edit form should have permissions you can't grant disabled, and the maximum priority set to your own, but that can be done in a subsequent PR, this one is big enough already

Gargron commented:
I'm not going to address localization simply because admins will not be able to translate their custom role titles into all languages anyway, by having custom role titles you give up the idea of localizing them. And I don't think keeping a set of translations for the default roles is worth it just for that initial creation.

Gargron committed:
Merge remote-tracking branch 'origin/main' into feature-dynamic-roles

Gargron committed:
Add migration for old settings and fix tootctl role management

Gargron commented:
- Changed automatically migrated admin role from `Owner` to `Admin`, real owner will need to use `tootctl`
- Added migration for `min_invite_role` and `show_staff_badge` settings
- Fixed `--role` option on `tootctl accounts create` and `tootctl accounts modify`

DaveT-T commented:
NOTE to people searching for solutions to USER ROLE Permissions issues.   Clarification can be found after reading this thread and referring to your /live/config/roles.yml file:

The new Role "Owner" is the superuser, what we thought or expect "Admin" would be.  To become what we expect to be "Admin" need to assign yourself as "Owner"

roles.yml shows
admin:
    position:  10
    permissions:
      (long list)
    
owner:
    name: Owner
    position:  1000
    permissions:
    - administrator

Users in this thread mention this in passing but language is esoteric and THEY know exactly what they mean but /we/  lack the context to understand their meaning.  Thanks all.

=======
Found this trying to explain why my Sidekiq view suddenly disappeared for all Admin users.    Needed to assign myself as "Owner" and then create a role to allow [ ] DevOps  and assign users to that role to view.

sidekiq requires   role&.can?(:view_devops)
